### ISSUE TITLE
wheel zoom blocked on zoom-out

### ISSUE THREAD

**Issue by kajocina** [bokeh/bokeh#6950]
I observe a problem with the wheel-zoom tool while using x_range and y_range bounds in figure call. 
The zoom-out function simply stops zooming out upon hitting only one of the defined boundaries. I would expect the zoom-out to bring back the figure to its initial state.

Using bokeh 0.12.9, Python 2.7.6, Ubuntu 14.04 LTS and updated Chrome/Firefox browsers.

Check the simple example code I attach:

```
from bokeh.models import Range1d, ColumnDataSource
from bokeh.plotting import figure, show, output_file


dataSrc = ColumnDataSource({'x':[-100,0,100],'y':[-100,0,100]})

fig = figure(tools=["wheel_zoom"],
             active_scroll="wheel_zoom",
             x_range=Range1d(-100, 100, bounds=(-100, 100)),
             y_range=Range1d(-100, 100, bounds=(-100, 100))
             )
fig.circle('x','y',source=dataSrc)
output_file("test_bug.html")
show(fig)
```

If you try zooming in the side of the the plot and from this state you move the cursor to the center and zoom-out, the zooming gets stuck. It will work correctly if you try zooming-out while pointing in the exact same spot.

Is it a known bug? Seems like a problem on Python side rather than something JS/browser related.

**Comment by bryevdv** [comment]
This was discussed at length in the past and it was decided this was the best possible behavior. If you allow the other dimension to zoom freely after another dimension has reached its bounds, then the point in location of the point under the cursor necessarily has to change to accommodate the situation. But the defining property of a wheel zoom (to me and others anyway) is that the zoom is always fixed exactly on the same point.

**Comment by kajocina** [comment]
I see your point albeit I feel that while this is not a problem for a person with knowledge about this module, it is rather confusing to end-users who don't even necessarily know that bokeh is creating plots for them in a web app they are using. 
Too bad it was decided to keep it this way ;-)

Best,
Piotr

**Comment by bryevdv** [comment]
@kajocina Bokeh is extensible (precisely because we will never be able to satisfy all the many conflicting needs/wants in the core library). If you want to implement a zoom tool with different behavior, it is a possibility:

https://bokeh.pydata.org/en/latest/docs/user_guide/extensions.html

**Comment by kajocina** [comment]
Thanks, I will sit down at this with my JS-specialist colleague and see
what can be done from our side.

Best,
Piotr

**Comment by bryevdv** [comment]
For reference the existing zoom tool implementation is here here:

https://github.com/bokeh/bokeh/blob/master/bokehjs/src/coffee/models/tools/gestures/wheel_zoom_tool.coffee

you could probably just subclass and do nothing but provide a specialized `_scroll` method to override the one there. It should call its own version of `scale_range`. The built in `scale_range` which is what checks range bounds, is here:

https://github.com/bokeh/bokeh/blob/master/bokehjs/src/coffee/core/util/zoom.coffee#L19

I suppose if you are interested in working up a PR I would consider adding a new property on wheel zoom tools to toggle this behavior (with the default being the current behavior)

**Comment by kernc** [comment]
Duplicate of https://github.com/bokeh/bokeh/issues/5821.
